nixos/spacecookie: convert into settings-style service module

sternenseemann opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

sternenseemann committed:
nixos/spacecookie: use nix style strings for description

sternenseemann committed:
nixos/tests/spacecookie: refactor

* Use proper gopher urls
* The client vms name is also controlled in a single place now
* fileContent holds the precise file content now
* wait for the spacecookie unit instead of the port
  * avoids sending an empty request
  * since spacecookie is a notify service it only is fully started when
    the socket has been set up.

aanderse commented on nixos/modules/services/networking/spacecookie.nix:
```diff
@@ -4,9 +4,14 @@ with lib;
-  });
+  } // cfg.extraConfig;
+
+  configFile = pkgs.writeText "spacecookie.json"
```
How about
```
format = pkgs.formats.json {};
configFile = formats.generate "spacecookie.json" cfg.settings;
```
and replace `extraConfig` with an option called `settings`, `type = formats.type;`.

sternenseemann replied:
The approach described in <https://github.com/NixOS/rfcs/pull/42> is actually a good fit for this module and the changes in the next release of `spacecookie`.

Is it possible to add optional options without defaults, so we'd still have documentation for options under `settings`?

aanderse replied:
Take a look in the NixOS manual for free form module options. I think you'll like what you see.

aanderse commented on nixos/modules/services/networking/spacecookie.nix:
```diff
@@ -18,19 +23,58 @@ in {
+        '';
+      };
+
+      package = mkOption {
```
Or you can use overlays. Either way works.

sternenseemann replied:
I think I want to keep this as an option here if somebody wants it as it is trivial to implement. Overriding a haskellPackage alone is confusing enough…

aanderse replied:
Sounds good 👍

aanderse reviewed (changes requested):
Just a few points. Let me know if you need any additional information, if you're not familiar with `pkgs.formats`, etc...

sternenseemann committed:
nixos/spacecookie: add package option

This allows to change the derivation to use for the spacecookie server
binary. We probably should also use justStaticExecutables by default to
reduce the runtime closure of the service.

sternenseemann committed:
nixos/spacecookie: add openFirewall option

Convenience shortcut which automatically configures the firewall to open
the port which is also configured for the spacecookie service.

sternenseemann committed:
nixos/spacecookie: add address option customizing listen address

This configuration option reflects a new feature from the unreleased
spacecookie version allowing to customize the address spacecookie will
listen on (e. g. "::1" to bind on link-local addresses only). We will
not use this feature in the future, since the configuration option of
spacecookie naturally only has an effect if we don't use socket
activation (and spacecookie sets up its own socket), but having the same
functionality in the service seems like a good idea.

We can luckily emulate this behavior with socket activation as well.

sternenseemann committed:
nixos/spacecookie: convert into settings-style freeform configuration

* Move `hostname` and `root` into a settings submodule with a freeform
  type, allowing users to also use options not known to the NixOS
  service. Compatibility with a warning for the renamed options is also
  trivial to achieve.
* `port` stays where it is as we don't actually use the `port` option of
  spacecookie to set up the socket, but only to inform spacecookie about
  the port we have set in the `systemd.socket` file, this makes more
  sense. Additionally the configuration of the listening port and
  address change in the next spacecookie release — we can dodge this
  issue altogether by doing our own thing, but I'm interested to hear
  opinions on this.
  To ensure that this is not misconfigured, we add an assertion for
  the port option.
* Add an assertion for `user` in settings which has no effect the way
  we are starting spacecookie as it wouldn't be able to call setuid.
  The message also explains how a specific user can be used with
  spacecookie if desired.

sternenseemann commented:
Okay, I have yet again refactored the service massively and am eager to hear feedback on it. Sorry, I took a bit, but since I was in the middle of making a new release for spacecookie anyways, I held back until I had everything sort of figured out (not released yet, but I know at least what the configuration changes will be).

New options:

* `package`, allowing to change the derivation
* `openFirewall`, doing the `allowedTCPPorts` dance for you
* `address`, allowing to customize the address to listen on

I also moved everything except `port` into the `settings` attribute which is converted into the config file using `pkgs.formats.json`.

The thing I'm a bit bothered about is the question of `port` and `address`. The problem here is that we don't actually use the capabilities of spacecookie for these options, as the socket is handled by systemd which is why I currently have it outside of `settings`, but this could be confusing for users possibly.

Another reason for this is, that in the next spacecookie release the top level `port` option will be deprecated in favor of something like this:

```json
{
  "listen": {
    "port" : 70,
    "addr" : "::1"
  }
}
```

…where `addr` does what `address` does in this version of the service, but won't have any effect if socket activation is used which we do (we only need to pass the port to spacecookie because this is relevant to know at runtime). I've thought about hijacking the `listen` set from `settings`, removing `addr` (because if set it generates a warning if socket activation is used) and using the information gathered to set up `spacecookie.socket`. However, the catch here is that the syntax for IPv6 addresses in `getaddrinfo(3)` and `systemd.socket(5)` is slightly different, so I feel like separating our socket configuration from the settings set makes things a bit clearer, but I'm interested to hear opinions on this.

I have the adjustments for the next version on a separate branch and will PR them when the update hits `haskellPackages`, but there's no rush or timing to consider as the configuration will stay compatible (even for `port`).

sternenseemann committed:
nixos/spacecookie: reflect changes for spacecookie 1.0.0.0

* New log options
* The old port option has been deprecated in favor of listen -> port

https://github.com/sternenseemann/spacecookie/blob/master/CHANGELOG.md#1000

aanderse commented:
@GrahamcOfBorg test spacecookie

aanderse commented:
@sternenseemann not sure what to say about that failed test... :man_shrugging:

sternenseemann commented:
> @sternenseemann not sure what to say about that failed test... man_shrugging

That's really weird, since the curl call seems to produce the correct output, but then fails with 56 (“Failure in receiving network data”). I'll ponder this some more tomorrow, I guess.

sternenseemann commented:
@ofborg test spacecookie

sternenseemann commented:
Seems like the `curl -f` call is flaky sometimes, I can't reproduce it locally and it doesn't happen on ofborg every time. Maybe we should remove the `succeed` and just check the output for curl to make sure we never block the channel accidentally?

sternenseemann commented:
Found a less ugly workaround for the curl problem, namely using netcat. Upstream issue is also fixed, really need to get this release out the door now :)

sternenseemann commented:
@ofborg test spacecookie

sternenseemann commented:
@ofborg test spacecookie

sternenseemann committed:
spacecookie: add top-level attribute for haskellPackages.spacecookie

The haskellPackages.spacecookie derivation also includes a library and
thus a lot of propagated haskell dependencies. The top-level attribute
uses haskell.lib.justStaticExecutables and therefore only the
executable. This should reduce the runtime closure users have to
download considerably if they only want the server.

sternenseemann commented:
With #116096 we have spacecookie 1.0.0.0 including a fix for the `curl` flakiness making the test fail! I've reverted the workaround for that in the test and added the new options' description to `settings`.

sternenseemann commented:
@ofborg test spacecookie

aanderse commented on nixos/modules/services/networking/spacecookie.nix:
```diff
@@ -4,43 +4,182 @@ with lib;
+          };
+
+          options.log = mkOption {
+            type = types.submodule {
```
You already applied a `freeformType` so I don't think you need to recursively do that.

sternenseemann replied:
You're right, thanks! The generated config file is a bit noisier now, but shouldn't really matter.

aanderse reviewed (approved):
Sorry this slipped off my radar... Are we good to merge? Are you planning on squashing any commits (I'm indifferent)?

sternenseemann commented:
Yes! Almost forgot it as well.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -18947,6 +18947,9 @@ in
 
   sogo = callPackage ../servers/web-apps/sogo { };
 
+  spacecookie =
+    haskell.lib.justStaticExecutables haskellPackages.spacecookie;
```
This newline is not required.

sternenseemann replied:
I don't understand why you keep pointing out formatting “issues” on already merged PRs. We have no established formatting guideline in the first place and especially with merged PRs this serves really nothing other than wasting the time of everyone involved.

SuperSandro2000 replied:
> I don't understand why you keep pointing out formatting “issues” on already merged PRs.

If I have 50 tabs open GitHub sometimes just forgets to refresh them and they are only updated after I posted a comment.

> We have no established formatting guideline in the first place

Just do reasonable formatting and that new line there is just because nixfmt enforces a 80 line length which we are not following in nixpkgs anyway.
